{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 10 Assignment\n",
    "\n",
    "**Due date**: Tue, Oct 27 by 3:30pm (submit on Sakai)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Building a bioinformatics pipeline, part I"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal of this assignment is to create the basis of a bioinformatics pipeline that can be used to study the variation and evolution of genes/proteins.  The pipeline you will build will accomplish the following steps:\n",
    "\n",
    "1. Read in a CSV delimited spreadsheet containing Genbank accession numbers\n",
    "\n",
    "2. Use Biopython to download to Genbank files corresponding to the accession numbers in step 1\n",
    "\n",
    "3. Extract the protein sequences of the genes represented in each genbank file, saving the results into a single FASTA file\n",
    "\n",
    "4. Align the sequences with the multiple sequence alignment program MUSCLE using the appropriate interface from Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data set of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use our pipeline to study interspecific variation in betacoronavirus spike proteins, and the protein that the spike proteins recognize and bind to on mammalian cells, called ACE2 (Angiotensin-Converting Enzyme 2). For details on background read the following papers:\n",
    "\n",
    "1. Hu, B., Guo, H., Zhou, P. et al. Characteristics of SARS-CoV-2 and COVID-19. Nat Rev Microbiol (2020). https://doi.org/10.1038/s41579-020-00459-7\n",
    "\n",
    "\n",
    "2. Broad and Differential Animal Angiotensin-Converting Enzyme 2 Receptor Usage by SARS-CoV-2\n",
    "Xuesen Zhao, Danying Chen, Robert Szabla, Mei Zheng, Guoli Li, Pengcheng Du, Shuangli Zheng, Xinglin Li, Chuan Song, Rui Li, Ju-Tao Guo, Murray Junop, Hui Zeng, Hanxin Lin\n",
    "Journal of Virology Aug 2020, 94 (18) e00940-20; https://doi.org/10.1128/JVI.00940-20\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problems"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Reading list of accessions [5 pts]\n",
    "\n",
    "Write a function to read a CSV formatted file that contains names and accession numbers for sequences of interest.  Each file will have two columns, \"Name\" and \"Accession\", giving the names of the organism that the sequence comes from and the corresponding Genbank accession numbers for each sequence. \n",
    "\n",
    "Your function should return a dictionary, with the names (strings) as the keys and the accession numbers (strings) as the values.\n",
    "\n",
    "The following files give accession list for a set of coronavirus spike proteins and a set of mammalian ACE2 genes:\n",
    "\n",
    "* `ACE2-accessions.csv`\n",
    "* `Spike-accessions.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_accessions(csvfile):\n",
    "    df = pd.read_csv(csvfile)\n",
    "    listName = df.Name.tolist()\n",
    "    listAccession = df.Accession.tolist()\n",
    "    retdict = dict(zip(listName,listAccession))\n",
    "    return retdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'SARS-CoV-2': 'MN908947', 'SARS-CoV BJ01': 'AY278488', 'RaTG13': 'MN996532.1', 'GX pangolin': 'MT040333.1', 'Bat WIV1': 'KF367457', 'Bat ZC45': 'MG772933.1', 'SARS-CoV Tor2': 'NC_004718.3'}\n",
      "{'Domestic dog': 'MT663955', 'Mexican free-tailed bat': 'MT663956', 'Chinese ferret badger': 'MT663957', 'Raccoon dog': 'MT663958', 'Domestic cat': 'MT663959', 'Rhesus monkey': 'MT663960', 'New Zealand White rabbit': 'MT663961', 'Hog badger': 'MT663962', 'Platypus': 'XM_001515547', 'Human': 'NM_001371415.1', 'Civet': 'AY881174.1', 'Rat': 'NM_001012006.1', 'Chinese horseshoe bat': 'KC881004.1', 'Pangolin': 'XM_017650263.1', 'Mouse': 'NM_001130513'}\n"
     ]
    }
   ],
   "source": [
    "# The following tests should work if your read_accessions function works correctly\n",
    "\n",
    "spikedict = read_accessions(\"Spike-accessions.csv\")\n",
    "assert(spikedict[\"RaTG13\"] == \"MN996532.1\")\n",
    "print(spikedict)\n",
    "\n",
    "ace2dict = read_accessions(\"ACE2-accessions.csv\")\n",
    "assert(ace2dict[\"Hog badger\"] == \"MT663962\")\n",
    "print(ace2dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Downloading Genbank files [7.5 pts]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Biopython includes facilities for automatically downloading Genbank files from the web, using NCBI's Entrez interface, as described in [Chap 5](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec59) and [Chap 9](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec%3Aefetch) of the Biopython tutorial.  \n",
    "\n",
    "**IMPORTANT** Before working on this problem make sure to read the [\"Entrez Guidelines\"](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec%3Aentrez-guidelines) about how to use the Entrez resources responsibly (if you don't do this you risk having your IP address banned from accessing Entrez). In particular make sure your specify your email address for validating Entrez access after importing the Etnrez library.\n",
    "\n",
    "``` \n",
    "from Bio import Entrez\n",
    "Entrez.email = \"A.N.Other@example.com\"  # replace with your Duke email\n",
    "```\n",
    "\n",
    "Write a function that takes as input an accession dictionary (e.g. created by the `read_accessions` function). For each accession, the function should use Biopython's Entrez tools to download the accessions of interest as genbank files and then each accession out to your file system, with file names formatted as `accession_number.gb`.  \n",
    "\n",
    "Your function should return a dictionary relating the names in the accession dictionary to the filenames of the genbank files.  For examp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_accessions(accessiondict):\n",
    "    #list of values\n",
    "    #access all values from Entrez\n",
    "    #make dictionary of values : file names as accession_number.gb\n",
    "    \n",
    "    fname = \"\"\n",
    "    listValueFiles = []\n",
    "    for i in accessiondict.values():\n",
    "        Entrez.email = \"ashrit.verma@duke.edu\"  # Always tell NCBI who you are\n",
    "        filename = i+'.gb'\n",
    "        listValueFiles.append(filename)\n",
    "        \n",
    "        if not os.path.isfile(filename):\n",
    "            # Downloading...\n",
    "            net_handle = Entrez.efetch(\n",
    "                db=\"nucleotide\", id=i, rettype=\"gb\", retmode=\"text\"\n",
    "            )\n",
    "            out_handle = open(filename, \"w\")\n",
    "            out_handle.write(net_handle.read())\n",
    "            out_handle.close()\n",
    "            net_handle.close()\n",
    "            print(\"Saved\")\n",
    "        print(\"Parsing...\")\n",
    "        record = SeqIO.read(filename, \"genbank\")\n",
    "        print(record)\n",
    "    \n",
    "    dictret = dict(zip(accessiondict.keys(),listValueFiles))\n",
    "    return dictret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing...\n",
      "ID: NM_001371415.1\n",
      "Name: NM_001371415\n",
      "Description: Homo sapiens angiotensin I converting enzyme 2 (ACE2), transcript variant 1, mRNA\n",
      "Number of features: 41\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=04-OCT-2020\n",
      "/accessions=['NM_001371415']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Angiotensin-converting enzymes (ACE, ACE2) gene variants and COVID-19 outcome', ...), Reference(title='Most frequent South Asian haplotypes of ACE2 share identity by descent with East Eurasian populations', ...), Reference(title='Analysis of ACE2 genetic variants in 131 Italian SARS-CoV-2-positive patients', ...), Reference(title='Mass Spectrometry and Structural Biology Techniques in the Studies on the Coronavirus-Receptor Interaction', ...), Reference(title='Cutaneous susceptibility to SARS-CoV-2 infection according to the expression of viral entry factors in the skin', ...), Reference(title='Angiotensin-converting enzyme 2 is an essential regulator of heart function', ...), Reference(title='Hydrolysis of biological peptides by human angiotensin-converting enzyme-related carboxypeptidase', ...), Reference(title='ACEH/ACE2 is a novel mammalian metallocarboxypeptidase and a homologue of angiotensin-converting enzyme insensitive to ACE inhibitors', ...), Reference(title='A human homolog of angiotensin-converting enzyme. Cloning and functional expression as a captopril-insensitive carboxypeptidase', ...), Reference(title='A novel angiotensin-converting enzyme-related carboxypeptidase (ACE2) converts angiotensin I to angiotensin 1-9', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from AC097625.11.\n",
      "This sequence is a reference standard in the RefSeqGene project.\n",
      "Summary: The protein encoded by this gene belongs to the\n",
      "angiotensin-converting enzyme family of dipeptidyl\n",
      "carboxydipeptidases and has considerable homology to human\n",
      "angiotensin 1 converting enzyme. This secreted protein catalyzes\n",
      "the cleavage of angiotensin I into angiotensin 1-9, and angiotensin\n",
      "II into the vasodilator angiotensin 1-7. ACE2 is known to be\n",
      "expressed in various human organs, and its organ- and cell-specific\n",
      "expression suggests that it may play a role in the regulation of\n",
      "cardiovascular and renal function, as well as fertility. In\n",
      "addition, the encoded protein is a functional receptor for the\n",
      "spike glycoprotein of the human coronavirus HCoV-NL63 and the human\n",
      "severe acute respiratory syndrome coronaviruses, SARS-CoV and\n",
      "SARS-CoV-2, the causative agent of coronavirus disease-2019\n",
      "(COVID-19). [provided by RefSeq, Aug 2020].\n",
      "Transcript Variant: This variant (1) encodes the longest isoform\n",
      "(1). Variants 1 and 2 encode the same isoform.\n",
      "Sequence Note: The RefSeq transcript and protein were derived from\n",
      "genomic sequence to make the sequence consistent with the reference\n",
      "genome assembly. The genomic coordinates used for the transcript\n",
      "record were based on alignments.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AB193260.1, AF241254.1 [ECO:0000332]')])), ('RefSeq-Attributes', OrderedDict([('coronavirus related', 'involved in SARS-CoV-2 infection'), ('MANE Ensembl match', 'ENST00000252519.8/ ENSP00000252519.3'), ('RefSeq Select criteria', 'based on manual assertion, conservation, expression, longest protein')]))])\n",
      "Seq('AGTCTAGGGAAAGTCATTCAGTGGATGTGATCTTGGCTCACAGGGGACGATGTC...CTC', IUPACAmbiguousDNA())\n",
      "{'Human': 'NM_001371415.1.gb'}\n"
     ]
    }
   ],
   "source": [
    "# test your function\n",
    "testdict = {\"Human\":\"NM_001371415.1\"}\n",
    "\n",
    "testfiledict = download_accessions(testdict)\n",
    "assert(testfiledict[\"Human\"] == \"NM_001371415.1.gb\")\n",
    "print(testfiledict)\n",
    "# Note that your test should also write a file with the name NM_001371415.1.gb in the directory in your working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Helper function to create Genbank files containing ONLY the spike protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_spike_cds(spikeaccessiondict):\n",
    "    newdict = {}\n",
    "    for key,value in spikeaccessiondict.items():\n",
    "        rec = SeqIO.read(value, format=\"gb\")\n",
    "        spikeftr = None\n",
    "        for ftr in rec.features:\n",
    "            if ftr.type == \"CDS\":\n",
    "                if \"gene\" in ftr.qualifiers:\n",
    "                    if \"S\" in ftr.qualifiers[\"gene\"]:\n",
    "                        spikeftr = ftr\n",
    "                        break\n",
    "                elif \"product\" in ftr.qualifiers:\n",
    "                    for descriptor in ftr.qualifiers[\"product\"]:\n",
    "                        if \"spike\" in descriptor:\n",
    "                            spikeftr = ftr\n",
    "                            break\n",
    "        if spikeftr is not None:\n",
    "            newrec = SeqRecord.SeqRecord(rec.seq, id=rec.id,name=rec.name,\n",
    "                                         description=rec.description, features=[spikeftr],\n",
    "                                         annotations=rec.annotations)\n",
    "            fname = rec.id + \"-spikeonly.gb\"\n",
    "            SeqIO.write(newrec, fname, format=\"gb\")\n",
    "            newdict[key] = fname\n",
    "    return newdict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Creating FASTA files [7.5 pts]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your next two functions will take the dictionary returned from the `download_accessions` function and then read each file, extracting either the nucleotide sequence or the  translations of the primary coding sequences (CDS) in each file, and will then write out a *single* FASTA file (`fastaname`) with the protein sequences for all the sequences of interest. After parsing the genbank files and writing the corresponding FASTA file, your functions should return the filename of the FASTA file it created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'allnucleotidesequences'"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def genbank2fasta_nucleotide(accessionfiledict, fastaname):\n",
    "    sequences = []\n",
    "    for key,value in accessionfiledict.items():\n",
    "        input_seq_iterator = SeqIO.parse(value, \"genbank\")\n",
    "        array1 = list(input_seq_iterator)\n",
    "        listftr = [ftr for ftr in array1[0].features if ftr.type == 'CDS']\n",
    "        if 0 < len(listftr):\n",
    "            f = listftr[0]\n",
    "            x = f.extract(array1[0].seq)\n",
    "            simple_seq = SeqRecord(x, id=value, name=key)\n",
    "            sequences.append(simple_seq)\n",
    "        savedir = \"C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/\" + fastaname\n",
    "        SeqIO.write(sequences, savedir, 'fasta')\n",
    "    return (fastaname)\n",
    "genbank2fasta_nucleotide(testfiledict, 'allnucleotidesequences')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'allproteinsequences'"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def genbank2fasta_protein(accessionfiledict, fastaname):\n",
    "    \n",
    "    #simple_seq is coding_dna\n",
    "    #coding_dna.translate()\n",
    "    sequences = []\n",
    "    for key,value in accessionfiledict.items():\n",
    "        input_seq_iterator = SeqIO.parse(value, \"genbank\")\n",
    "        array1 = list(input_seq_iterator)\n",
    "        listftr = [ftr for ftr in array1[0].features if ftr.type == 'CDS']\n",
    "        if 0 < len(listftr):\n",
    "            f = listftr[0]\n",
    "            x = f.extract(array1[0].seq)\n",
    "            simple_seq = SeqRecord(x, id=value, name=key)\n",
    "            translated_seq = simple_seq.translate()\n",
    "            sequences.append(translated_seq)\n",
    "        savedir = \"C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/\" + fastaname\n",
    "        SeqIO.write(sequences, savedir, 'fasta')\n",
    "    return (fastaname)\n",
    "genbank2fasta_protein(testfiledict, 'allproteinsequences')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Aligning protein sequences using MUSCLE [5 pts]\n",
    "\n",
    "[MUSCLE](http://www.drive5.com/muscle/) is a popular multiple sequence alignment program.  You can download a command line version of MUSCLE at that link, and \"install\" it by unzipping the binary and placing it somewhere in your path (I will show you how to do this in class).\n",
    "\n",
    "Once MUSCLE is on your computer you should test MUSCLE \"by hand\" (we'll do this in class).  Biopython also provides an interface to interact with MUSCLE from within a Python pipeline, as described in [Section 6.5](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec92) of the Biopython tutorial.\n",
    "\n",
    "Using the `MuscleCommandLine` interface that Biopython provides, write a function that aligns the amino acid sequences in a FASTA file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from io import StringIO\n",
    "from Bio import AlignIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_via_MUSCLE(infastafile, outfastafile):\n",
    "    muscle_exe = \"C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/muscle3.8.31.exe\"\n",
    "    in_file = \"C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/\" + infastafile\n",
    "    out_file = \"C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/\" + outfastafile\n",
    "    cline = MuscleCommandline(muscle_exe, input=in_file,out=out_file)\n",
    "    cline()\n",
    "    return(cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MuscleCommandline(cmd='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/muscle3.8.31.exe', input='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/allproteinsequences', out='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/Test-MUSCLE-Output.fasta')"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align_via_MUSCLE(genbank2fasta_protein(testfiledict, 'allproteinsequences'), \"Test-MUSCLE-Output.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Test the entire pipeline [5 pts]\n",
    "\n",
    "If each of your functions above works, then the following small code block implements the entire pipeline from start to finish:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing...\n",
      "ID: MN908947.3\n",
      "Name: MN908947\n",
      "Description: Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 23\n",
      "/molecule_type=ss-RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=18-MAR-2020\n",
      "/accessions=['MN908947']\n",
      "/sequence_version=3\n",
      "/keywords=['']\n",
      "/source=Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)\n",
      "/organism=Severe acute respiratory syndrome coronavirus 2\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='A new coronavirus associated with human respiratory disease in China', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=On Jan 17, 2020 this sequence version replaced MN908947.2.\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'Megahit v. V1.1.3'), ('Sequencing Technology', 'Illumina')]))])\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: AY278488.2\n",
      "Name: AY278488\n",
      "Description: SARS coronavirus BJ01, complete genome\n",
      "Number of features: 12\n",
      "/molecule_type=RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=01-SEP-2009\n",
      "/accessions=['AY278488']\n",
      "/sequence_version=2\n",
      "/keywords=['']\n",
      "/source=SARS coronavirus BJ01\n",
      "/organism=SARS coronavirus BJ01\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='The E protein is a multifunctional membrane protein of SARS-CoV', ...), Reference(title='SARS coronavirus BJ01 isolate genome sequence', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=On May 1, 2003 this sequence version replaced AY278488.1.\n",
      "Seq('CCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACT...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MN996532.1\n",
      "Name: MN996532\n",
      "Description: Bat coronavirus RaTG13, complete genome\n",
      "Number of features: 21\n",
      "/molecule_type=RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=24-MAR-2020\n",
      "/accessions=['MN996532']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Bat coronavirus RaTG13\n",
      "/organism=Bat coronavirus RaTG13\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='A pneumonia outbreak associated with a new coronavirus of probable bat origin', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=[WARNING] On Oct 13, 2020 this sequence was replaced by MN996532.2.\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'CLC Genomics Workbench v. 12.0'), ('Sequencing Technology', 'Illumina')]))])\n",
      "Seq('CTTTCCAGGTAACAAACCAACGAACTCTCGATCTCTTGTAGATCTGTTCTCTAA...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT040333.1\n",
      "Name: MT040333\n",
      "Description: Pangolin coronavirus isolate PCoV_GX-P4L, complete genome\n",
      "Number of features: 10\n",
      "/molecule_type=RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=23-APR-2020\n",
      "/accessions=['MT040333']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Pangolin coronavirus\n",
      "/organism=Pangolin coronavirus\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus', 'unclassified Sarbecovirus']\n",
      "/references=[Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'CLC Genomic Workbench v. v.9.0'), ('Sequencing Technology', 'Illumina')]))])\n",
      "Seq('GCTTTAACTTTTTACAAATCCCAGGTAGCAAAACCAACCAACTCTCGATCTCTT...TTC', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: KF367457.1\n",
      "Name: KF367457\n",
      "Description: Bat SARS-like coronavirus WIV1, complete genome\n",
      "Number of features: 14\n",
      "/molecule_type=ss-RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=06-NOV-2013\n",
      "/accessions=['KF367457']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Bat SARS-like coronavirus WIV1\n",
      "/organism=Bat SARS-like coronavirus WIV1\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='Isolation and characterization of a bat SARS-like coronavirus that uses the ACE2 receptor', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'Newbler v. 2.6'), ('Coverage', '>300x'), ('Sequencing Technology', 'Sanger dideoxy sequencing; 454')]))])\n",
      "Seq('ATATTAGGTTTTTACCTACCCAGGAAAAGCCAACCAACCCCGATCTCTTGTAGA...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MG772933.1\n",
      "Name: MG772933\n",
      "Description: Bat SARS-like coronavirus isolate bat-SL-CoVZC45, complete genome\n",
      "Number of features: 12\n",
      "/molecule_type=RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=29-SEP-2020\n",
      "/accessions=['MG772933']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Bat SARS-like coronavirus\n",
      "/organism=Bat SARS-like coronavirus\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='Genomic characterization and infectivity of a novel SARS-like coronavirus in Chinese bats', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATATTAGGTTTTTACCTTCCCAGGTAACAAACCAACTAACTCTCGATCTCTTGT...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: NC_004718.3\n",
      "Name: NC_004718\n",
      "Description: SARS coronavirus Tor2, complete genome\n",
      "Database cross-references: BioProject:PRJNA485481\n",
      "Number of features: 45\n",
      "/molecule_type=RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=15-SEP-2020\n",
      "/accessions=['NC_004718']\n",
      "/sequence_version=3\n",
      "/keywords=['RefSeq']\n",
      "/source=SARS coronavirus Tor2\n",
      "/organism=SARS coronavirus Tor2\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='Analysis of multimerization of the SARS coronavirus nucleocapsid protein', ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence is identical to AY274119.\n",
      "On or before Mar 28, 2016 this sequence version replaced\n",
      "NC_028858.1, NC_028866.1, NC_028884.1, NC_028893.1, NC_028873.1,\n",
      "NC_028845.1, NC_009696.1, NC_009695.1, NC_013664.1, NC_009693.1,\n",
      "NC_009694.1, NC_004718.2.\n",
      "Annotation based on that found in PMID: 31987001, PMID: 31967327\n",
      "and the annotation of P0C6X7.1.\n",
      "COMPLETENESS: full length.\n",
      "Seq('ATATTAGGTTTTTACCTACCCAGGAAAAGCCAACCAACCTCGATCTCTTGTAGA...AAA', IUPACAmbiguousDNA())\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MuscleCommandline(cmd='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/muscle3.8.31.exe', input='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/Spike-prot.fasta', out='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/Spike-prot-aligned.fasta')"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spikedict = read_accessions(\"Spike-accessions.csv\")\n",
    "spikefiledict = download_accessions(spikedict)\n",
    "spikeonlydict = filter_spike_cds(spikefiledict)\n",
    "spikenucfasta = genbank2fasta_nucleotide(spikeonlydict, \"Spike-nucs.fasta\")\n",
    "spikeprotfasta = genbank2fasta_protein(spikeonlydict, \"Spike-prot.fasta\")\n",
    "align_via_MUSCLE(spikeprotfasta, \"Spike-prot-aligned.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing...\n",
      "ID: MT663955.1\n",
      "Name: MT663955\n",
      "Description: Canis lupus familiaris angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663955']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Canis lupus familiaris (dog)\n",
      "/organism=Canis lupus familiaris\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Caniformia', 'Canidae', 'Canis']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGCTCTTCCTGGCTCCTTCTCAGCCTTGCTGCTCTAACTGCTGCTCAA...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663956.1\n",
      "Name: MT663956\n",
      "Description: Tadarida brasiliensis angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663956']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Tadarida brasiliensis (Brazilian free-tailed bat)\n",
      "/organism=Tadarida brasiliensis\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Chiroptera', 'Microchiroptera', 'Molossidae', 'Tadarida']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTTGGGCTCATTCTGGCTCCTTCTCAGCCTTGTTGCTGTGACTGCTGCTCAA...TGA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663957.1\n",
      "Name: MT663957\n",
      "Description: Melogale moschata angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663957']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Melogale moschata (Chinese ferret-badger)\n",
      "/organism=Melogale moschata\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Caniformia', 'Mustelidae', 'Helictidinae', 'Melogale']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTTAGGCTCTTCCTGGCTCCTTCTCAGCCTCGCTGCTCTAACTGCTGCTCAG...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663958.1\n",
      "Name: MT663958\n",
      "Description: Nyctereutes procyonoides angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663958']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Nyctereutes procyonoides (raccoon dog)\n",
      "/organism=Nyctereutes procyonoides\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Caniformia', 'Canidae', 'Nyctereutes']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGCTCTTCCTGGCTCCTTCTCAGCCTTGCTGCTCTAACTGCTGCTCAA...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663959.1\n",
      "Name: MT663959\n",
      "Description: Felis catus angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663959']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Felis catus (domestic cat)\n",
      "/organism=Felis catus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Feliformia', 'Felidae', 'Felinae', 'Felis']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGCTCTTTCTGGCTCCTTCTCAGCTTTGCTGCTTTAACTGCTGCTCAA...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663960.1\n",
      "Name: MT663960\n",
      "Description: Macaca mulatta angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663960']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Macaca mulatta (Rhesus monkey)\n",
      "/organism=Macaca mulatta\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Cercopithecidae', 'Cercopithecinae', 'Macaca']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGCTCTTCCTGGCTCCTTCTCAGCCTTGTTGCTGTAACTGCTGCTCAG...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663961.1\n",
      "Name: MT663961\n",
      "Description: Oryctolagus cuniculus angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663961']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Oryctolagus cuniculus (rabbit)\n",
      "/organism=Oryctolagus cuniculus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Glires', 'Lagomorpha', 'Leporidae', 'Oryctolagus']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGTTCTTCCTGGCTCCTTCTCAGCCTTGTGGCTGTCACTGCTGCTCAG...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: MT663962.1\n",
      "Name: MT663962\n",
      "Description: Arctonyx collaris angiotensin I converting enzyme 2 mRNA, complete cds\n",
      "Number of features: 2\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=10-JUL-2020\n",
      "/accessions=['MT663962']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Arctonyx collaris\n",
      "/organism=Arctonyx collaris\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Caniformia', 'Mustelidae', 'Melinae', 'Arctonyx']\n",
      "/references=[Reference(title='Broad and differential animal ACE2 receptor usage by SARS-CoV-2', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTTAGGCTCTTCCTGGCTCCTTCTCAGCCTCGCTGCTCTAACTGCTGCTCAG...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: XM_001515547.4\n",
      "Name: XM_001515547\n",
      "Description: PREDICTED: Ornithorhynchus anatinus angiotensin I converting enzyme 2 (ACE2), mRNA\n",
      "Database cross-references: BioProject:PRJNA534073\n",
      "Number of features: 3\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=29-APR-2019\n",
      "/accessions=['XM_001515547']\n",
      "/sequence_version=4\n",
      "/keywords=['RefSeq']\n",
      "/source=Ornithorhynchus anatinus (platypus)\n",
      "/organism=Ornithorhynchus anatinus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Monotremata', 'Ornithorhynchidae', 'Ornithorhynchus']\n",
      "/comment=MODEL REFSEQ:  This record is predicted by automated computational\n",
      "analysis. This record is derived from a genomic sequence\n",
      "(NC_041742.1) annotated using gene prediction method: Gnomon.\n",
      "Also see:\n",
      "    Documentation of NCBI's Annotation Process\n",
      "On Apr 29, 2019 this sequence version replaced XM_001515547.3.\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Full annotation'), ('Annotation Name', 'Ornithorhynchus anatinus Annotation Release 104'), ('Annotation Version', '104'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation pipeline'), ('Annotation Software Version', '8.2'), ('Annotation Method', 'Best-placed RefSeq; Gnomon'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('TCAGTGCTCAGGGCCCAACCCAAGTTCAAAGGCTGCAGGAAGAGACAAATCCAT...AAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: NM_001371415.1\n",
      "Name: NM_001371415\n",
      "Description: Homo sapiens angiotensin I converting enzyme 2 (ACE2), transcript variant 1, mRNA\n",
      "Number of features: 41\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=04-OCT-2020\n",
      "/accessions=['NM_001371415']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Angiotensin-converting enzymes (ACE, ACE2) gene variants and COVID-19 outcome', ...), Reference(title='Most frequent South Asian haplotypes of ACE2 share identity by descent with East Eurasian populations', ...), Reference(title='Analysis of ACE2 genetic variants in 131 Italian SARS-CoV-2-positive patients', ...), Reference(title='Mass Spectrometry and Structural Biology Techniques in the Studies on the Coronavirus-Receptor Interaction', ...), Reference(title='Cutaneous susceptibility to SARS-CoV-2 infection according to the expression of viral entry factors in the skin', ...), Reference(title='Angiotensin-converting enzyme 2 is an essential regulator of heart function', ...), Reference(title='Hydrolysis of biological peptides by human angiotensin-converting enzyme-related carboxypeptidase', ...), Reference(title='ACEH/ACE2 is a novel mammalian metallocarboxypeptidase and a homologue of angiotensin-converting enzyme insensitive to ACE inhibitors', ...), Reference(title='A human homolog of angiotensin-converting enzyme. Cloning and functional expression as a captopril-insensitive carboxypeptidase', ...), Reference(title='A novel angiotensin-converting enzyme-related carboxypeptidase (ACE2) converts angiotensin I to angiotensin 1-9', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from AC097625.11.\n",
      "This sequence is a reference standard in the RefSeqGene project.\n",
      "Summary: The protein encoded by this gene belongs to the\n",
      "angiotensin-converting enzyme family of dipeptidyl\n",
      "carboxydipeptidases and has considerable homology to human\n",
      "angiotensin 1 converting enzyme. This secreted protein catalyzes\n",
      "the cleavage of angiotensin I into angiotensin 1-9, and angiotensin\n",
      "II into the vasodilator angiotensin 1-7. ACE2 is known to be\n",
      "expressed in various human organs, and its organ- and cell-specific\n",
      "expression suggests that it may play a role in the regulation of\n",
      "cardiovascular and renal function, as well as fertility. In\n",
      "addition, the encoded protein is a functional receptor for the\n",
      "spike glycoprotein of the human coronavirus HCoV-NL63 and the human\n",
      "severe acute respiratory syndrome coronaviruses, SARS-CoV and\n",
      "SARS-CoV-2, the causative agent of coronavirus disease-2019\n",
      "(COVID-19). [provided by RefSeq, Aug 2020].\n",
      "Transcript Variant: This variant (1) encodes the longest isoform\n",
      "(1). Variants 1 and 2 encode the same isoform.\n",
      "Sequence Note: The RefSeq transcript and protein were derived from\n",
      "genomic sequence to make the sequence consistent with the reference\n",
      "genome assembly. The genomic coordinates used for the transcript\n",
      "record were based on alignments.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AB193260.1, AF241254.1 [ECO:0000332]')])), ('RefSeq-Attributes', OrderedDict([('coronavirus related', 'involved in SARS-CoV-2 infection'), ('MANE Ensembl match', 'ENST00000252519.8/ ENSP00000252519.3'), ('RefSeq Select criteria', 'based on manual assertion, conservation, expression, longest protein')]))])\n",
      "Seq('AGTCTAGGGAAAGTCATTCAGTGGATGTGATCTTGGCTCACAGGGGACGATGTC...CTC', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: AY881174.1\n",
      "Name: AY881174\n",
      "Description: Paguma larvata angiotensin-converting enzyme 2 (ACE2) mRNA, complete cds\n",
      "Number of features: 3\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=11-APR-2005\n",
      "/accessions=['AY881174']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Paguma larvata (masked palm civet)\n",
      "/organism=Paguma larvata\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Carnivora', 'Feliformia', 'Viverridae', 'Paradoxurinae', 'Paguma']\n",
      "/references=[Reference(title='Isolation and characterization of viruses related to the SARS coronavirus from animals in southern China', ...), Reference(title='Angiotensin-converting enzyme 2 is a functional receptor for the SARS coronavirus', ...), Reference(title='Receptor and viral determinants of SARS-coronavirus adaptation to human ACE2', ...), Reference(title='Direct Submission', ...)]\n",
      "Seq('ATGTCAGGCTCTTTCTGGCTCCTTCTCAGCTTTGCTGCTTTAACTGCTGCTCAA...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: NM_001012006.1\n",
      "Name: NM_001012006\n",
      "Description: Rattus norvegicus angiotensin I converting enzyme 2 (Ace2), mRNA\n",
      "Number of features: 17\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=ROD\n",
      "/date=12-OCT-2020\n",
      "/accessions=['NM_001012006', 'XM_228924']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=Rattus norvegicus (Norway rat)\n",
      "/organism=Rattus norvegicus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Glires', 'Rodentia', 'Myomorpha', 'Muroidea', 'Muridae', 'Murinae', 'Rattus']\n",
      "/references=[Reference(title='A SARS-CoV-2 Infection Model in Mice Demonstrates Protection by Neutralizing Antibodies', ...), Reference(title='Altered COVID-19 receptor ACE2 expression in a higher risk group for cerebrovascular disease and ischemic stroke', ...), Reference(title='The pathogenicity of SARS-CoV-2 in hACE2 transgenic mice', ...), Reference(title=\"The host's angiotensin-converting enzyme polymorphism may explain epidemiological findings in COVID-19 infections\", ...), Reference(title='Dynamic Variation of RAS on Silicotic Fibrosis Pathogenesis in Rats', ...), Reference(title='Characterization of renal angiotensin-converting enzyme 2 in diabetic nephropathy', ...), Reference(title='Angiotensin-converting enzyme 2 is an essential regulator of heart function', ...), Reference(title='Hydrolysis of biological peptides by human angiotensin-converting enzyme-related carboxypeptidase', ...), Reference(title='A human homolog of angiotensin-converting enzyme. Cloning and functional expression as a captopril-insensitive carboxypeptidase', ...), Reference(title='A novel angiotensin-converting enzyme-related carboxypeptidase (ACE2) converts angiotensin I to angiotensin 1-9', ...)]\n",
      "/comment=PROVISIONAL REFSEQ: This record has not yet been subject to final\n",
      "NCBI review. The reference sequence was derived from AY881244.1.\n",
      "On Feb 11, 2005 this sequence version replaced XM_228924.2.\n",
      "Summary: enzyme peptidyl-dipeptidase; mouse homolog has a role in\n",
      "regulation of cardiac contractility and regulating Angiotensin 1\n",
      "and 2 levels [RGD, Feb 2006].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AY881244.1, GQ262788.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns SAMEA2689597, SAMEA2689599 [ECO:0000348]')]))])\n",
      "Seq('ATGTCAAGCTCCTGCTGGCTCCTTCTCAGCCTTGTTGCTGTTGCTACTGCTCAG...TAA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: KC881004.1\n",
      "Name: KC881004\n",
      "Description: Rhinolophus sinicus isolate Rs-3357 angiotensin-converting enzyme 2 (ACE2) mRNA, complete cds\n",
      "Number of features: 3\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=22-NOV-2013\n",
      "/accessions=['KC881004']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Rhinolophus sinicus (Chinese rufous horseshoe bat)\n",
      "/organism=Rhinolophus sinicus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Chiroptera', 'Microchiroptera', 'Rhinolophidae', 'Rhinolophinae', 'Rhinolophus']\n",
      "/references=[Reference(title='Isolation and characterization of a bat SARS-like coronavirus that uses the ACE2 receptor', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n",
      "Seq('ATGTCAGGCTCTTCCTGGCTCCTTCTGAGCCTTGTTGCTGTAACCACTGCTCAA...TAG', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: XM_017650263.1\n",
      "Name: XM_017650263\n",
      "Description: PREDICTED: Manis javanica angiotensin I converting enzyme 2 (ACE2), transcript variant X2, mRNA\n",
      "Database cross-references: BioProject:PRJNA335369\n",
      "Number of features: 3\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=MAM\n",
      "/date=29-JUL-2016\n",
      "/accessions=['XM_017650263']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=Manis javanica (Malayan pangolin)\n",
      "/organism=Manis javanica\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Laurasiatheria', 'Pholidota', 'Manidae', 'Manis']\n",
      "/comment=MODEL REFSEQ:  This record is predicted by automated computational\n",
      "analysis. This record is derived from a genomic sequence\n",
      "(NW_016527422.1) annotated using gene prediction method: Gnomon.\n",
      "Also see:\n",
      "    Documentation of NCBI's Annotation Process\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Full annotation'), ('Annotation Version', 'Manis javanica Annotation Release 100'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation pipeline'), ('Annotation Software Version', '7.1'), ('Annotation Method', 'Best-placed RefSeq; Gnomon'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('CCAGGATGAACTTTATATTGGTTCAGCAGATTGTTTACTGTTTTTTCTCCTTCT...TTA', IUPACAmbiguousDNA())\n",
      "Parsing...\n",
      "ID: NM_001130513.1\n",
      "Name: NM_001130513\n",
      "Description: Mus musculus angiotensin I converting enzyme (peptidyl-dipeptidase A) 2 (Ace2), transcript variant 1, mRNA\n",
      "Number of features: 33\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=ROD\n",
      "/date=21-OCT-2020\n",
      "/accessions=['NM_001130513']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=Mus musculus (house mouse)\n",
      "/organism=Mus musculus\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Glires', 'Rodentia', 'Myomorpha', 'Muroidea', 'Muridae', 'Murinae', 'Mus', 'Mus']\n",
      "/references=[Reference(title='Mouse model of SARS-CoV-2 reveals inflammatory role of type I interferon signaling', ...), Reference(title='Animal and translational models of SARS-CoV-2 infection and COVID-19', ...), Reference(title='Animal models for COVID-19', ...), Reference(title='Animal Models to Study Emerging Technologies Against SARS-CoV-2', ...), Reference(title='The ACE2-deficient mouse: A model for a cytokine storm-driven inflammation', ...), Reference(title='Efficient replication of severe acute respiratory syndrome coronavirus in mouse cells is limited by murine angiotensin-converting enzyme 2', ...), Reference(title='Mechanisms of host defense following severe acute respiratory syndrome-coronavirus (SARS-CoV) pulmonary infection of mice', ...), Reference(title='Heart block, ventricular tachycardia, and sudden death in ACE2 transgenic mice with downregulated connexins', ...), Reference(title='Molecular cloning, mRNA expression and chromosomal localization of mouse angiotensin-converting enzyme-related carboxypeptidase (mACE2)', ...), Reference(title='Angiotensin-converting enzyme 2 is an essential regulator of heart function', ...)]\n",
      "/comment=VALIDATED REFSEQ: This record has undergone validation or\n",
      "preliminary review. The reference sequence was derived from\n",
      "BY722909.1 and AK141604.1.\n",
      "Transcript Variant: This variant (1) represents the longer\n",
      "transcript.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'EF408739.1, EF408740.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns SAMN00849387, SAMN01164133 [ECO:0000348]')]))])\n",
      "Seq('AGGCCCATGAGCCCTGCCATTTAAAGTGGCTCCTCTCTTACACTCTGGGAATGA...TCT', IUPACAmbiguousDNA())\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MuscleCommandline(cmd='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/muscle3.8.31.exe', input='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/Ace2-prot.fasta', out='C:/Users/cleve/OneDrive/Documents/Notes/Notebooks Linked Materials/Duke University/Bio 208 Computing on the Genome/Assignment Week 10/Ace2-prot-aligned.fasta')"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ace2dict = read_accessions(\"ACE2-accessions.csv\")\n",
    "ace2filedict = download_accessions(ace2dict)\n",
    "ace2nucfasta = genbank2fasta_nucleotide(ace2filedict, \"Ace2-nucs.fasta\")\n",
    "ace2protfasta = genbank2fasta_protein(ace2filedict, \"Ace2-prot.fasta\")\n",
    "align_via_MUSCLE(ace2protfasta, \"Ace2-prot-aligned.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
